src/uu/pgrep/src/pgrep.rs

```diff
@@ -0,0 +1,278 @@
+    let result = || {
+        if matches.get_flag("count") {
+            return format!("{}", result.len());
+        };
+
+        let delimiter = matches.get_one::<String>("delimiter").unwrap();
+
+        let formatted: Vec<_> = if matches.get_flag("list-full") {
+            result
+                .into_iter()
+                .map(|it| format!("{} {}", it.pid, it.cmdline))
+                .collect()
+        } else if matches.get_flag("list-name") {
+            result
+                .into_iter()
+                .map(|it| format!("{} {}", it.pid, it.clone().status().get("Name").unwrap()))
+                .collect()
+        } else {
+            result.into_iter().map(|it| format!("{}", it.pid)).collect()
+        };
+
+        formatted.join(delimiter)
+    };
+
+    let result = result();
+    if !result.is_empty() {
+        println!("{}", result);
+    };
+
+    Ok(())
+}
+
+fn collect_arg_patterns(matches: &ArgMatches) -> Vec<String> {
+    let should_ignore_case = matches.get_flag("ignore-case");
+
+    let programs = matches
+        .get_many::<String>("pattern")
+        .unwrap_or_default()
+        .map(|it| it.to_string());
+
+    if should_ignore_case {
+        programs.map(|it| it.to_lowercase()).collect::<Vec<_>>()
+    } else {
+        programs.collect()
+    }
+}
+
+fn collect_pid(matches: &ArgMatches, patterns: &[String]) -> Vec<PidEntry> {
+    let should_inverse = matches.get_flag("inverse");
+    let should_ignore_case = matches.get_flag("ignore-case");
+
+    let flag_full = matches.get_flag("full");
+    let flag_exact = matches.get_flag("exact");
+
+    let predicate_closure = |mut it: PidEntry| {
+        let binding = it.status();
+        let name = binding.get("Name")?;
+
+        // Processs flag `--ignore-case`
+        let name = if should_ignore_case {
+            name.to_lowercase()
+        } else {
+            name.into()
+        };
+
+        let mut iter = patterns.iter();
+
+        let name_matched = if flag_full {
+            // Equals `Name` in /proc/<pid>/status
+            iter.any(|it| name.eq(it))
+        } else if flag_exact {
+            // Equals `cmdline` in /proc/<pid>/cmdline
+            iter.any(|s| it.cmdline.eq(s))
+        } else {
+            // Contains `cmdline` in /proc/<pid>/cmdline
+            iter.any(|it| name.contains(it))
+        };
+
+        let tty_matched = if let Some(ttys) = matches.get_many::<String>("terminal") {
+            // convert from input like `pts/0`
+            let ttys = ttys
+                .cloned()
+                .flat_map(TerminalType::try_from)
+                .collect::<HashSet<_>>();
+
+            if let Ok(value) = it.ttys() {
+                value.iter().any(|it| ttys.contains(it))
+            } else {
+                false
+            }
+        } else {
+            true
+        };
+
+        if (name_matched || tty_matched) ^ should_inverse {
+            Some(it)
+        } else {
+            None
+        }
```

Comment:
Returning an `Option` is odd. A predicate usually returns a `bool`.

Reply:
I renamed this it to `evaluate` because `it` needs to be mutable within itself (which have side effects). Using a bool as a return value could cause performance degradation.